test/lib/jdk/test/lib/serial/SerialObjectBuilder.java

```diff
@@ -0,0 +1,217 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.test.lib.serial;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.UncheckedIOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import static java.io.ObjectStreamConstants.*;
+
+/**
+ * A basic builder of a serial object.
+ */
+public class SerialObjectBuilder {
+
+    private final ObjectOutputStream objectOutputStream;
+    private final ByteArrayOutputStream byteArrayOutputStream;
+
+    private record NameAndType<T>(String name, Class<T>type) { }
+
+    private String className;
+    private long suid;
+    private SerialObjectBuilder superClass;
+    private final LinkedHashMap<NameAndType<?>, Object> primFields = new LinkedHashMap<>();
+    private final LinkedHashMap<NameAndType<?>, Object> objectFields = new LinkedHashMap<>();
+
+    private SerialObjectBuilder() {
+        try {
+            byteArrayOutputStream = new ByteArrayOutputStream();
+            objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public static SerialObjectBuilder newBuilder(String className) {
+        return (new SerialObjectBuilder()).className(className);
+    }
+
+    private SerialObjectBuilder className(String className) {
+        this.className = className;
+        return this;
+    }
+
+    public SerialObjectBuilder suid(long suid) {
+        this.suid = suid;
+        return this;
+    }
+
+    public SerialObjectBuilder superClass(SerialObjectBuilder superClass) {
+        this.superClass = superClass;
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addPrimitiveField(String name, Class<T> type, T value) {
+        if (!type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected non-primitive field: " + type);
+        primFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    public <T> SerialObjectBuilder addField(String name, Class<T> type, T value) {
+        if (type.isPrimitive())
+            throw new IllegalArgumentException("Unexpected primitive field: " + type);
+        objectFields.put(new NameAndType<>(name, type), value);
+        return this;
+    }
+
+    private static int getPrimitiveSignature(Class<?> cl) {
```

Comment:
The new java.lang.constant.TypeDescriptor will provide the primitive type from the Class.
Class implements TypeDescriptor including descriptorString.

This method can be replaced with clazz.descriptorString().